src/pages/tutorial/cw-contract/query.mdx

```diff
@@ -0,0 +1,323 @@
+[`Deserialize`](https://docs.serde.rs/serde/trait.Deserialize.html) traits from `serde` crate.
+
+Then we need to implement our entry point. It is very similar to the `instantiate` one. The first
+significant difference is a type of `deps` argument. For `instantiate`, it was a `DepMut`, but here
+we went with a [`Deps`](https://docs.rs/cosmwasm-std/latest/cosmwasm_std/struct.Deps.html) object.
+That is because the query can never alter the smart contract's internal state. It can only read the
+state. It comes with some consequences - for example, it is impossible to implement caching for
+future queries (as it would require some data cache to write to).
+
+The other difference is the lack of the `info` argument. The reason here is that the entry point
+which performs actions (like instantiation or execution) can differ how an action is performed based
+on the message metadata - for example, they can limit who can perform an action (and do so by
+checking the message `sender`). It is not a case for queries. Queries are supposed just purely to
+return some transformed contract state. It can be calculated based on some chain metadata (so the
+state can "automatically" change after some time), but not on message info.
+
+Note that our entry point still has the same `Empty` type for its `msg` argument - it means that the
+query message we would send to the contract is still an empty JSON: `{}`
+
+The last thing that changed is the return type. Instead of returning the `Response` type on the
+success case, we return an arbitrary serializable object. This is because queries are not using a
+typical actor model message flow - they cannot trigger any actions nor communicate with other
+contracts in ways different than querying them (which is handled by the `deps` argument). The query
+always returns plain data, which should be presented directly to the querier.
+
+Now take a look at the implementation. Nothing complicated happens there - we create an object we
+want to return and encode it to the
+[`Binary`](https://docs.rs/cosmwasm-std/latest/cosmwasm_std/struct.Binary.html) type using the
+[`to_json_binary`](https://docs.rs/cosmwasm-std/latest/cosmwasm_std/fn.to_json_binary.html)
+function.
+
+## Improving the message
+
+We have a query entry point, but there is a problem with the query message. It is always an empty
+JSON. It is terrible - if we would like to add another query in the future, it would be difficult to
+have any reasonable distinction between query variants.
+
+In practice, we address this by using a non-empty query message type. Improve our contract:
+
+```rust {11-14, 27, 30-38} copy filename="src/lib.rs" template="empty"
+use cosmwasm_std::{
+    entry_point, to_json_binary, Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+struct QueryResp {
+    message: String,
+}
+
+#[derive(Serialize, Deserialize)]
```

Comment:
Our convention is to encode enum variants as snake_case JSON names. We either use the [`rename_all` Serde attribute](https://docs.rs/cosmwasm-schema/latest/cosmwasm_schema/attr.cw_serde.html) or go through the `cw_serde` macro.

Reply:
Same as for the `cw_serde` comment. Since we introduce concept of JSON schema in later pages the idea is to keep new information to minimum.